### PR TITLE
Adds tests for search api methods responses Fixes #1418

### LINKED ISSUE #1418: Add more functional tests to each method in Search API
Hi, this is a [first-timers-only issue](https://github.com/publiclab/plots2/projects/2). This means we've worked to make it more legible to folks who either **haven't contributed to our codebase before, or even folks who haven't contributed to open source before**. 

If that's you, we're interested in helping you take the first step and can answer questions and help you out as you do. Note that we're especially interested in contributions from people from groups underrepresented in free and open source software!

If you have contributed before, consider leaving this one for someone new, and looking through our general [help wanted](https://github.com/publiclab/plots2/labels/help-wanted) issues. Thanks!

## The problem

> Tests help us ensure that the site behaves as we expect it to -- and that as new features are added, they don't break existing functions. We're working to get to near 100% "coverage" -- which would mean all our code is covered by tests.

We have good tests written for the typeahead API, here: https://github.com/publiclab/plots2/blob/master/test/functional/typeahead_api_test.rb

This tests each method in: https://github.com/publiclab/plots2/blob/master/app/api/srch/typeahead.rb

But we should have similar ones for the search API, for **each of** the methods in: https://github.com/publiclab/plots2/blob/master/app/api/srch/search.rb

(Note that these draw heavily on this file, in turn -- in case you want to know how they're generated: https://github.com/publiclab/plots2/blob/master/app/services/search_service.rb)

## Solution

In #1409, a first-time contributor recently created a new test file, and a single test, which you can find at `test/functional/search_api_test.rb`: 

```ruby
  test 'search notes functionality' do
    get '/api/srch/notes?srchString=Blog'
    assert last_response.ok?
    # more stuff after this
```

This is a basic test that the response is "ok" and then it goes on to test what's in the response. See how it uses a URL to access the `notes` method on this line: 

So, copy it, and follow the provided URL examples in `search.rb` to create a new test for a different method. For example, `profiles`: https://github.com/publiclab/plots2/blob/master/app/api/srch/search.rb#L30

Start by just asserting `last_response.ok?` as above; we can add more substantive tests later.

You can try running this with `rake test` or (just to run the functional tests only) `rake test TEST=test/functional/search_api_test.rb` -- hopefully it will pass! 

But you can also just open a pull request (see below) and our automated test system will run it, and you'll be able to see the output there. 

If you can help with this, please open a pull request with this test added, and let's see how it goes -- even if it doesn't pass, we can help you with the next steps.

## Steps to Fix

- [ ] claim this issue with a comment here, below, and ask any clarifying questions you need
- [ ] set up a repository locally following the README instructions, and make sure that all tests pass
- [ ] try to fix the issue following the steps above, but even before you're done, you can:
- [ ] commit your changes and start a pull request (see [contributing to Public Lab software](https://publiclab.org/wiki/contributing-to-public-lab-software)) but mark it as "in progress" if you have questions or if you haven't finished
- [ ] alert someone via the developers list (see below) to have your pull request merged. This may take a few extra steps depending on when you do it, but we'll help you out!

Please email the developers list (see https://publiclab.org/wiki/developers) or go to [the chatroom](http://publiclab.org/chat) if you have questions, and take a look at [our first-timers landing page for more information](https://github.com/publiclab/plots2/projects/2)!

### PR DESCRIPTION
**Pull request by mlendale**
Fixes #1418

---
This change is [Reviewable](https://reviewable.io/reviews/publiclab/plots2/1437)

### CODE REVIEW

File: test/functional/search_api_test.rb [hunk 1]
```diff
@@ -26,7 +50,67 @@ def app
...
+    pattern = {
+      srchParams: {
+        srchString: 'Blog',
+        seq: nil,
+      }.ignore_extra_keys!
+    }.ignore_extra_keys!
+
+    matcher = JsonExpressions::Matcher.new(pattern)
+
+    json = JSON.parse(last_response.body)
```
Comment (jywarren) [inline test/functional/search_api_test.rb]: Hi, here, we'd want to search for a profile, not "Blog" - so we could change `get '/api/srch/profiles?srchString=Blog'` to `get '/api/srch/profiles?srchString=Bob' (working from the test data in `/test/fixtures/users.yml`). 

Then, as done above on [these lines](https://github.com/publiclab/plots2/pull/1437/files#diff-ae2647e0aa24f1474d9192f6cedc811eR27), we can test the returned data. In this case, we'd want to check that the returned data is the record for "Bob". We could make an educated guess but maybe the best would be to start by inserting a logging action to see what the contents are:

```
puts json['items'][0]
```

Once we know exactly what it returns, we can adjust the test to confirm that. Does this make sense?

File: Gemfile.lock [hunk 2]
```diff
@@ -1,31 +1,31 @@
 GEM
   remote: https://rubygems.org/
   specs:
-    actionmailer (3.2.22.4)
-      actionpack (= 3.2.22.4)
+    actionmailer (3.2.22.5)
```
Comment (jywarren) [inline Gemfile.lock]: So, here, we don't really want to submit a new `Gemfile.lock` -- do you think you could remove that file from the commit using this guide? https://stackoverflow.com/questions/215718/reset-or-revert-a-specific-file-to-a-specific-revision-using-git

**Comment by david-days** [comment]
@jywarren, This kind of work could fall under the planning issue I just put in to day, Issue #1449 if you want an overarching guide to this kind of work.

@mlendale, if you enjoy the work on the RESTful services and @jywarren agrees,, I'd like to invite you to keep adding more work under Issue #1449.  

For Phase 1, I'm adding the Swagger UI to the plots2 code as an initial documentation, testing, and learning tool.  Then it's more of the work you've done here.

**Comment by jywarren** [comment]
Hi, just a check in if you could use any help with this one!

**Comment by mlendale** [comment]
I definitely missed your comments. They all make sense, of course. Doing it
ASAP. Sorry for the delay.

2017-06-13 0:24 GMT+02:00 Jeffrey Warren <notifications@github.com>:

**Comment by jywarren** [comment]
No problem at all! Thanks and no huge rush but we're here to help if you
have questions.

File: test/functional/search_api_test.rb [hunk 3]
```diff
@@ -26,7 +50,67 @@ def app
...
+  end
+
+  test 'search questions functionality' do
+    get '/api/srch/questions?srchString=Blog'
+    assert last_response.ok?
+
+    # Expected search pattern
+    pattern = {
+      srchParams: {
+        srchString: 'Blog',
```
Comment (jywarren) [inline test/functional/search_api_test.rb]: And here, there's probably not a question with the search string "Blog" -- but check the fixtures data in `/test/fixtures/nodes.yml` and see what search term would be appropriate for a questions search. Thanks!

File: test/functional/search_api_test.rb [hunk 4]
```diff
@@ -26,7 +50,67 @@ def app
...
+  end
+
+  test 'search tags functionality' do
+    get '/api/srch/tags?srchString=Blog'
+    assert last_response.ok?
+
+    # Expected search pattern
+    pattern = {
+      srchParams: {
+        srchString: 'Blog',
```
Comment (jywarren) [inline test/functional/search_api_test.rb]: And here, check out `/test/fixtures/tags.yml` for a tag to search for, so that it returns at least one result.

**Comment by jywarren** [comment]
Hi! I think I'm going to close this one up as we haven't heard back -- but we're still very interested in helping you get involved if you'd like! We have a new welcoming page and a whole lot of good issues up for grabs -- and ready to offer guidance if you need it. Thank you!!

https://publiclab.github.io/community-toolbox/#r=all

**Pull request closed without merging**